src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -3841,18 +3976,23 @@ IEnumerator IEnumerable.GetEnumerator()
 
         public IGrouping<TKey, TElement>[] ToArray()
         {
-            IArrayProvider<IGrouping<TKey, TElement>> lookup = Lookup<TKey, TElement>.Create<TSource>(_source, _keySelector, _elementSelector, _comparer);
+            IIListProvider<IGrouping<TKey, TElement>> lookup = Lookup<TKey, TElement>.Create<TSource>(_source, _keySelector, _elementSelector, _comparer);
             return lookup.ToArray();
         }
 
         public List<IGrouping<TKey, TElement>> ToList()
         {
-            IListProvider<IGrouping<TKey, TElement>> lookup = Lookup<TKey, TElement>.Create<TSource>(_source, _keySelector, _elementSelector, _comparer);
+            IIListProvider<IGrouping<TKey, TElement>> lookup = Lookup<TKey, TElement>.Create<TSource>(_source, _keySelector, _elementSelector, _comparer);
             return lookup.ToList();
         }
+
+        public int Count
+        {
+            get { return Lookup<TKey, TElement>.Create<TSource>(_source, _keySelector, _elementSelector, _comparer).Count; }
+        }
```

Comment:
Not yet. Currently this only gets hit by `Count()` . I've an unpushed (and not yet pushable) branch that does hit this for precisely that sort of optimisation, but that also introduces a `ConstantTimeCount` property.